src/ffi.rs

```diff
@@ -363,6 +363,38 @@ unsafe fn strlen(mut str_ptr: *const c_char) -> usize {
 }
 
 
+pub(crate) static EMPTY: [u8; 1] = [0];
+/// A trait for producing pointers that will always be valid in C.
+/// Rust doesn't promise what pointers does it give to ZST (https://doc.rust-lang.org/nomicon/exotic-sizes.html#zero-sized-types-zsts)
+/// This trait will give a pointer to a static variable in case the type is empty.
+pub(crate) trait CPtr {
+    type Target;
+    fn as_c_ptr(&self) -> *const Self::Target;
+    fn as_mut_c_ptr(&mut self) -> *mut Self::Target;
+}
+
+impl<T> CPtr for [T] {
+    type Target = T;
+    fn as_c_ptr(&self) -> *const Self::Target {
+        if self.is_empty() {
+            EMPTY.as_ptr() as *const _
```

Comment:
Isn't `ptr::null()` better here and more expected/idiomatic?

The issue with `EMPTY` is that it is a valid pointer. The C code typically checks for NULL pointer to avoid dereferencing it. But here the pointer will not be NULL and worse, the object is only 1 byte long. So in the worst , case, the C code could try to read say 65 bytes from it.

One thing we need to be careful with are cases where pointer can be NULL and this gives us difference semantics, e.g., https://github.com/bitcoin-core/secp256k1/blob/master/include/secp256k1.h#L477. Then we need to be sure that we don't confuse the empty slice with the NULL pointer. The surrounding code should handle this already now in all such cases (e.g., by length checks etc) but we should double-check.

Reply:
the NULL checks in C call the callback function, so this will cause a panic.

If you try to parse an empty slice as a public key you should get an error, not a panic

Reply:
But my point is that the surrounding code needs to handle these case properly then. secp256k1-rust is a wrapper around secp256k1, so if the wrapper allows for arbitrary-size slices (instead of fixed size as in the wrapped C), then it needs to check the size and return a proper error if it's not 33 bytes.

Reply:
Maybe that's gonna be the proper fix? not enabling arbitrary-size slices? and instead allowing only `[u8;64]` or `[u8;33]`?

What you're proposing sounds more like #144

Reply:
Hm, yes. :D Right, we can enforce this statically, so I guess we should do it. @apoelstra What do you think?

Reply:
Hmm, it does require the user to create a fixed-size array and use `copy_from_slice` which is a bit unergonomic. But the current situation requires the user call `unwrap` themselves.

Alternately we could check the length and panic in our code, similar to what `copy_from_slice` does.

FWIW the current API predates `copy_from_slice` which was only introduced in Rust 1.9.

Reply:
no matter what these functions will require a `Result` and an `unwrap/handling` on the user side, because the pubkey or privkey can be incorrect (zero/above n/not a valid point)

Reply:
FWIW the `ThirtyTwoByteHash` trait was supposed to fix this. But it does not because neither `bitcoin_hashes` nor `secp256k1` want to depend on each other..

Reply:
Maybe we should have a `from_slice` which panics on any form of invalid key and a `try_from_slice` which returns an option?

Reply:
> Maybe we should have a `from_slice` which panics on any form of invalid key and a `try_from_slice` which returns an option?

Together with methods for fixed array sizes, this sounds reasonable: Then we just have both: nice fixed arrays if you can use it and if you just have a slice, you don't need to call `copy_from_slice`...

Reply:
@real-or-random how many functions are you talking about? 3 different ones? (33 bytes, 64 bytes, and the try one) should both the 33 and 64 panic if key is invalid?